File: Binary Tree Level Order Traversal II/Binary Tree Level Order Traversal II.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > levelOrderBottom(TreeNode *root) {
        std::vector<std::vector<int> >  res;
        
        if(!root)   return res;
        
        std::vector<TreeNode*>    curr_level;
        curr_level.push_back(root);
        
        while(!curr_level.empty()){
            std::vector<int>    container;
            std::vector<TreeNode*>  next_level;
            for(unsigned int i = 0; i < curr_level.size(); ++i){
                container.push_back(curr_level[i]->val);
                if(curr_level[i]->left)             next_level.push_back(curr_level[i]->left);
                if(curr_level[i]->right)       next_level.push_back(curr_level[i]->right);
            }
            res.push_back(container);
            curr_level = next_level;
        }
        
        for(unsigned int i = 0; i < res.size() / 2; ++i){
            std::vector<int>    tmp = res[i];
            res[i] = res[res.size() - 1 - i];
            res[res.size() - 1 - i] = tmp;
        }
        
        return res;
    }
};

File: Binary Tree Level Order Traversal II/Binary Tree Level Order Traversal II.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> bfs = new ArrayList<List<Integer>>();
        if(root == null){
            return bfs;
        }
        
        List<TreeNode> workingList = new ArrayList<TreeNode>();
        workingList.add(root);
        List<Integer> rootval = new ArrayList<Integer>();
        rootval.add(root.val);
        bfs.add(0, rootval);
        while(!workingList.isEmpty()){
            List<TreeNode> newList = new ArrayList<TreeNode>();
            List<Integer> values = new ArrayList<Integer>();
            for(TreeNode n : workingList){
                if(n.left != null){
                    newList.add(n.left);
                    values.add(n.left.val);
                }
                if(n.right != null){
                    newList.add(n.right);
                    values.add(n.right.val);
                }
            }
            workingList = newList;
            if(!values.isEmpty()){
                bfs.add(0, values);
            }
        }
        
        return bfs;
    }
}
